ENH: Reduce memory usage of arraysetops._unique1d

klaussfreire opened the pull request:
The current implementation creates lots of temporal copies of the
input array, which results in unnecessarily inflated memory
requirements. The case is particularly bad when invoked with
return_index = True.

As an example, `numpy.unique(a, return_index = True)` where a.size = 200M 
and a.dtype = int64 will use about 10GB of memory, where the input 
is merely 1.6GB. That's a 600% overhead for an operation that should be
doable with far less than that.

This patch tries to reduce memory wastage by releasing temporary
copies as soon as we're done with them. It also introduces a return_data,
default True, to be able to skip returning the actual unique data and thus
avoid one whole extra copy if all you want are the indices, and a return_mask,
default False, to be able to obtain the unique mask when interesting.

It also introduces assume_sorted and sort_inplace (default False) flags
to handle presorted input in the first case, and discardable input in the
second case (ie: if all you want is to reduce an array to their unique values
and don't care about the original array afterwards).

This allows the low-overhead call forms like 
`a = numpy.unique(a, sort_inplace = True)`, which will do
the sort in-place and only allocate an array of masks temporarily.

klaussfreire committed:
Reduce memory usage of arraysetops.unique1d

The current implementation creates lots of temporal copies of the
input array, which results in unnecessarily inflated memory
requirements. The case is particularly bad when invoked with
return_index = True.

This patch tries to reduce memory wastage by releasing temporary
copies as soon as we're done with them.

klaussfreire committed:
FEATURE: Add return_flags, retun_data to numpy.unique

Make returning actual array data an optional thing in
numpy.unique and unique1d, to make it possible to compute indexes
or flags with minimal memory overhead (returning a copy of the
data requires memory, wasted if the caller isn't interested in that)

Add an option to return the flags, rather than the indexes,
which may be enough for some uses.

klaussfreire committed:
DOC: Add docstring for new unique args

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -160,6 +168,9 @@ def unique(ar, return_index=False, return_inverse=False,
         .. versionadded:: 1.9.0
+    unique_flags : ndarray, optional
+        The flags of the first occurrences of the unique values in the
+        original array. Only provided if `return_flags` is True.
```
`mask` would be a much better name than `flags` here, I think. (and in `return_flags`, obviously)

I see you chose it as it matched the internal `flag` variable, but I don't think that was a good name either.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -240,58 +262,77 @@ def reshape_uniq(uniq):
         else:
-            ret = (ar,)
+            if return_data:
+                ret = (ar,)
```
This would be cleaner as another `+=` like the others, and starting with `ret = ()` - even if it's (negligeably) slower

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -240,58 +262,77 @@ def reshape_uniq(uniq):
             if return_counts:
                 ret += (np.empty(0, np.intp),)
+            if return_flags:
+                ret += (np.empty(0, np.bool8),)
```
This is an unusual spelling of `np.bool_`

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -240,58 +262,77 @@ def reshape_uniq(uniq):
+                        or return_flags or not return_data)
 
     if ar.size == 0:
         if not optional_returns:
```
I think this would be better handled by just building the `ret` tuple, and unpacking it if it has length 1 - since now all the results are optional

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -193,6 +204,17 @@ def unique(ar, return_index=False, return_inverse=False,
+    Return a mask of the original array that give the unique values:
+
+    >>> a = np.array(['a', 'b', 'b', 'c', 'a'])
+    >>> indices = np.unique(a, return_flags=True, return_data=False)
```
Doesn't this return a 1-tuple with this implementation? Please add some tests!

klaussfreire replied:
It does. I was under the impression this was a doctest. Will fix.

eric-wieser reviewed (commented):
I'm +1 on allowing `unique` to just return a mask or indices, but I think this needs some more work first. In particular, it needs tests!

klaussfreire commented:
> I'm +1 on allowing unique to just return a mask or indices, but I think this needs some more work first. In particular, it needs tests!

I was under the impression the docstring examples were being run already (ie: as doctests)

klaussfreire committed:
MAINT: Refactor unique to return tuple only if necessary

Make unique return a tuple only when more than one value is
requested. Also rename flag to mask, which is a better match
for what it is.

klaussfreire committed:
ENH: unique: add assume_sorted and sort_inplace

Add tests too. Tests revealed an issue with the returned masks,
that they were relative to the sorted array, and not the input
array, so they were hard to use as intended.

The intended use case was for presorted input, so implemented
flags that optimize unique for presorted input or for input that
can be sorted in-place.

klaussfreire commented:
> I'm +1 on allowing unique to just return a mask or indices, but I think this needs some more work first. In particular, it needs tests!

Ok, after adding tests and noticing an issue with the returned masks, it became clear that the PR was a tad simplistic for the intended use case. Namely, to be able to quickly de-duplicate array of structured items when the input needs not be preserved (ie: the original array can be clobbered).

The masks were intended to be used as:

``` python
a = numpy.array([1,2,3,4,1,2,3,4])
mask = np.unique(a, return_mask=True, return_data=False)
a = a[mask]
```

However, some tests exposed the issue that the returned masks are applicable to the sorted input, not the actual input.

This last push adds 2 more flags to np.unique to better support that use case (and other similar cases): `assume_sorted` and `sort_inplace`. The use case above can now be written as:

``` python
a = np.unique(a, sort_inplace=True)
```

The other flags remain since they might empower other use cases.

One optimization made is when doing unique over multidimensional arrays that need to be copied to reshape them or make them contiguous. In that case, sort_inplace will be implicitly set and a whole copy will be avoided, which seems like a good thing.

eric-wieser commented:
It seems more sensible to me to just have the user call `a.sort()` and specify `assume_sorted`, rather than have a `sort_inplace` flag.

klaussfreire commented:
> It seems more sensible to me to just have the user call a.sort() and specify assume_sorted, rather than have a sort_inplace flag.

Ok

klaussfreire committed:
MAINT: Remove sort_inplace flag

klaussfreire committed:
TST: Add test for multidimensional masks

klaussfreire committed:
TST: Add multidimensional return_data=False tests

klaussfreire committed:
BUG: Fix reshaping on return_data=False

klaussfreire commented:
> > It seems more sensible to me to just have the user call a.sort() and specify assume_sorted, rather than have a sort_inplace flag.
> Ok

I started doing that, but now I'm not so sure it's wise.

Doing the tests for multidimentional matrices, it's evident the sort call isn't at all trivial for some data types and matrix shapes. Having the caller do all that contortion seems to defeat the point of calling np.unique.

So I'd rather keep that flag.

klaussfreire committed:
MAINT: Reintroduce sort_inplace, add tests

The sort for multidimensional matrices is far from trivial, so
simplify the caller's life with sort_inplace. Add tests for
the sort_inplace interface for multidimensional matrices.

klaussfreire committed:
TST: Add sort_inplace tests for 1d arrays

klaussfreire committed:
MAINT: Fix mask_is_sorted for some corner cases

klaussfreire committed:
MAINT: Fix typo

klaussfreire committed:
DOC: Clarify that sort_inplace may not always be honored

klaussfreire commented:
No comments?

eric-wieser commented:
> it's evident the sort call isn't at all trivial for some data types and matrix shapes.

You're right, I'd forgotten about this

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -111,15 +111,20 @@ def ediff1d(ary, to_end=None, to_begin=None):
-    comes up in the input array.
+    reconstruct the input array, the number of times each unique value
+    comes up in the input array, and the mask of the input array that gives
+    the unique values.
```
This might work better as a bulleted list.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +295,124 @@ def reshape_uniq(uniq):
+    orig_base = consolidated
+    while (orig_base.base is not None and orig_base is not orig_ar
+           and orig_base.base is not orig_ar):
+        orig_base = orig_base.base
```
What exactly are you trying to do here? I think this deserves some comments, and the `orig_base is orig_ar or orig_base.base is orig_ar` code from below calculated alongside it.

eric-wieser replied:
Or better, a helper function

klaussfreire replied:
Trying to figure out whether `consolidated` is a view whose underlying data points to `orig_ar` even indirectly. I found no helper function for that. I could add one I guess.

klaussfreire replied:
Done, added `np.array_view_root` with corresponding tests.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +295,124 @@ def reshape_uniq(uniq):
+        # Will need the permutation to compute masks
+        optional_indices = True
+    if ((not optional_indices and not assume_sorted and not sort_inplace)
+        or len(ar.shape) != 1):
```
This is better spelt `ar.ndim`, and same below

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +295,124 @@ def reshape_uniq(uniq):
+        # Otherwise, we don't need to make a copy
+        if sort_inplace and len(ar.shape) == 2 and ar.shape[1] == 1:
+            # consolidated (n x 1) matrix, no need to copy
+            ar = ar.reshape(ar.size)
```
Better spelt `.ravel()`

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +295,124 @@ def reshape_uniq(uniq):
+        if return_mask:
+            ret += (np.empty(0, np.bool_),)
+        if len(ret) == 1:
+            ret = ret[0]
```
Why not move this outside `_unique1d` into `unique`?

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -270,27 +270,30 @@ def _unique1d(ar, return_index=False, return_inverse=False,
+            inv_idx = np.empty(shape, dtype=np.intp)
             inv_idx[perm] = iflag
             ret += (inv_idx,)
+        perm = None
```
Why not `del perm`?

klaussfreire replied:
Leftover from a previous version, fixing

klaussfreire committed:
DOC: Turn optional outputs into a bulleted list

klaussfreire committed:
MAINT: simplify code using ravel/ndim where useful

klaussfreire committed:
MAINT: Simplify unique1d by always returning a tuple

Plus some minor fixes

klaussfreire committed:
ENH: Add array_view_root utility function to core

Adds array_view_root utility to follow views to their
root underlying data, useful to figure out if modifying an
array will touch another's data

eric-wieser reviewed (commented):
Darn, this was sitting around...

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +300,119 @@ def reshape_uniq(uniq):
+    if return_index or return_inverse:
+        # We won't perform an in-place sort even if requested
+        mask_is_sorted = assume_sorted
+    elif orig_base is consolidated_base:
```
This part of the change makes me uneasy - can we leave out this `.base` traversal to avoid a copy to a different PR? It would be nice to get `return_mask` into `1.14.

klaussfreire replied:
This part is key though, without it, copies still have to be made, and that increases memory usage many-fold

klaussfreire replied:
Would it be ok to make it opt-in? (say, `try_inplace = True`)

eric-wieser replied:
I'd still rather separate the new feature (`return_mask`) and the `.base`-checking optimization into separate PRs - I'd want a second opinion on the latter half, which will be easier to get if the diff is smaller

klaussfreire commented on numpy/lib/arraysetops.py:
```diff
@@ -239,59 +295,124 @@ def reshape_uniq(uniq):
+            inv_idx = np.empty(shape, dtype=np.intp)
+            inv_idx[perm] = imask
             ret += (inv_idx,)
+        perm = None
```
`perm` may not be initialized at this point, and `del perm` would throw an error in that case

klaussfreire committed:
Merge branch 'master' of github.com:numpy/numpy

eric-wieser commented:
In the interest of moving forward with this, I've extracted some of your simpler changes and some extra cleanup to #10498. Once those are in, then this'll need rebasing - something I can try to do if needed.

klaussfreire commented:
Sorry about the slowness, I intended to do this, but I've been backed up with other work.

klaussfreire committed:
Merge branch 'master' of github.com:numpy/numpy

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,119 @@ def reshape_uniq(uniq):
+
+    if not optional_returns:
+        ret = (ar[mask],)
+    else:
```
No need for this `if` any more

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,119 @@ def reshape_uniq(uniq):
     optional_indices = return_index or return_inverse
-
+    optional_returns = (optional_indices or return_counts
+                        or return_mask or not return_data)
```
This variable can be eliminated too

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,119 @@ def reshape_uniq(uniq):
+            ret += (np.empty(0, np.intp),)
+        if return_mask:
+            ret += (np.empty(0, np.bool_),)
+        return ret
```
I don't think this is needed either

klaussfreire replied:
Yep, sorry, I botched the merge. Fixed now.

klaussfreire committed:
MAINT: Re-apply #10498 after merge

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,100 @@ def reshape_uniq(uniq):
 
     optional_indices = return_index or return_inverse
 
+    if (not optional_indices and not assume_sorted
```
No need to condition on `optional_indices` here

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,100 @@ def reshape_uniq(uniq):
+                    optional_indices = True
+
+    shape = ar.shape
+    size = ar.size
```
No need for these any more, just use `mask.shape` and `mask.size`, like was done before

klaussfreire committed:
MAINT: Remove unneded shape/size locals

klaussfreire committed:
MAINT: Remove unneeded conditional on optinal_indices

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +310,98 @@ def reshape_uniq(uniq):
+    orig_base = np.array_view_root(orig_ar)
+    if return_index or return_inverse:
+        # We won't perform an in-place sort even if requested
+        mask_is_sorted = assume_sorted
```
Should this set `sort_inplace = False`?

klaussfreire replied:
Almost, if we sort in-place oportunistically on a copy, `mask_is_sorted != assume_sorted or sort_inplace`. But some simplification of the sort is doable, just pushed that.

klaussfreire committed:
MAINT: simplify sort_inplace/mask_is_sorted logic

klaussfreire committed:
MAINT: Reflow to avoid 80 chars limit

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +316,94 @@ def reshape_uniq(uniq):
+    if not indirect_sort and orig_base is not consolidated_base:
+        # We're operating on a copy, so we might as well
+        # do an in-place sort on that copy
+        sort_inplace = True
```
Why not:
```
if return_index or return_inverse:
    sort_inplace = False
else if orig_base is not consolidated_base:
    sort_inplace = True
```
and eliminate the stuff below.

AT the very least, this would be much clearer as an `else` on the `if indirect_sort`

klaussfreire replied:
This is pretty much that, except the first part is at the beginning of the function so the 1d case also gets the `sort_inplace = False` bit.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -254,42 +316,94 @@ def reshape_uniq(uniq):
+        sort_inplace = True
+
+    # The mask is sorted iff the input ends up sorted
+    mask_is_sorted = assume_sorted or (sort_inplace and orig_base is consolidated_base)
```
Why is `orig_base is consolidated_base` needed here?

klaussfreire replied:
`numpy/lib/tests/test_arraysetops.py:TestUnique.test_unique_axis` trips without it.

That test performs an in-place sort on a temporary copy (ie: sort_inplace was forced to True). So, sort_inplace doesn't imply mask_is_sorted if the sort is done on a copy. Only when the sort is done on the input array (ie: when the input array ends up sorted) mask_is_sorted.

I tried to clarify that on the comment above.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -217,10 +254,35 @@ def unique(ar, return_index=False, return_inverse=False,
+    indirect_sort = return_index or return_inverse
+    if indirect_sort:
+        # We won't perform an in-place sort even if requested
+        sort_inplace = False
```
Why not do one, if it was requested explictly?

klaussfreire replied:
`return_index` and `return_inverse` both require an argsort, not an inplace sort. Getting the original array in sorted order while still computing the indexes or the inverse is really cumbersome and inefficient, and not all that useful. So they weren't implemented.

Basically, sort_inplace is more a permission to do the sort in place than a request. It means: it's ok to modify the original sort order.

eric-wieser committed:
Merge branch 'master' into master

eric-wieser commented:
Went through my old comments and marked most of them as resolved (thanks!), and solved the conflicts here. I'm not sure this is ready to merge yet though.

I also have a hunch that `array_view_root(a)` is exactly equal to `a.base or a`, ie the base attribute already points all the way to the bottom.

seberg commented:
Yes, `array_view_root` should be exactly equivalent to just using `a if a.base is None else a.base`. We make sure of that during base setting.

klaussfreire commented:
> Yes, `array_view_root` should be exactly equivalent to just using `a if a.base is None else a.base`. We make sure of that during base setting.

That's not true for masked arrays:

``` pycon
>>> from numpy import ndarray                                                                                                                                                                                                                                                  
>>> from numpy.ma import masked_array                                                                                                                                                                                                                                          
>>> mx1 = masked_array([[1, 2], [3, 4]], mask=[[True, False], [False, True]])                                                                                                                                                                                                  
>>> mx2 = mx1.ravel()                                                                                                                                                                                                                                                          
>>> mx1.base                                                                                                                                                                                                                                                                   
array([[1, 2],                                                                                                                                                                                                                                                                 
       [3, 4]])                                                                                                                                                                                                                                                                
>>> mx2.base                                                                                                                                                                                                                                                                   
array([1, 2, 3, 4])                                                                                                                                                                                                                                                            
```

With python 3.6 and numpy 1.16.4

klaussfreire committed:
ENH: Fix docstring in array_view_root

Clarify it takes an array_like

klaussfreire committed:
BUG: Fix unique dispatcher to match unique's new signature

klaussfreire committed:
BUG: Fix array_view_root's module attribute

seberg commented:
Thanks for opening the PR!  Considering the age, I am going to close it at this point, I would be more than happy about a new similar PR, though!

Right now, I also don't think the approach is good enough, however.  The input could be an array-like where it is currently not easy to tell whether or not a copy was made (right now I am not sure it is even possible, it may require deeper additions to the array creation code).

Some of the other improvements are likely good on first glance, and it would be cool to rescue them, however the code is changed a fair bit, since the PR unfortunately.

A new PR based on this work against the `main` branch would be cool but is unfortunately probably required to revive this.
